### ISSUE TITLE
Audit and handle remaining cases of `rustc::potential_query_instability` lint (iterating `HashMap`s)

### ISSUE THREAD

**Issue by Aaron1011** [rust-lang/rust#84447]
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/issues/84447)*

The iteration order of a `HashMap` depends on the `Hash` value of the keys. In rustc, the `Hash` value of a key can change between compilation sessions, even if the `HashStable` value remains the same (e.g. `DefId`, `Symbol`, etc.). This means that iterating over a `HashMap` can cause a query to produce different results given the same (as determined by `HashStable`) input. This can be quite subtle, as demonstrated by https://github.com/rust-lang/rust/pull/82272#issuecomment-825124318

To eliminate this source of issues, we should make an internal rustc-only lint for iterating over a `HashMap` in any way. This will need to cover explicit method calls (e.g. `.keys()` and `.values()`), as well as any usage of the `IntoIterator` impl for `HashMap`.

Note that we don't want to ban the usage of `HashMap`s with unstable keys - as long as they're only used for lookup (and not iteration), it's much more efficient than repeatedly converting to the stable form of a key (e.g. `DefId` -> `DefPathHash` or `Symbol` -> `String`).

**Comment by bjorn3** [comment]
> Note that we don't want to ban the usage of HashMaps with unstable keys - as long as they're only used for lookup (and not iteration),

https://github.com/rust-lang/rust/pull/64131 added the `StableMap` wrapper for `FxHashMap` that doesn't expose the iteration order.

**Comment by jyn514** [comment]
@bjorn3 see https://github.com/rust-lang/rust/issues/84446

**Comment by cjgillot** [comment]
The lint exists since https://github.com/rust-lang/rust/pull/89558, and has allowed on all compiler crates as `rustc::potential_query_instability`.
We now need to stop allowing it in the compiler.

Steps, for each crate except `rustc_data_structures`:
- remove the `allow(rustc::potential_query_instability)` in `lib.rs`;
- replace the `FxHash{Map,Set}` by `FxIndex{Map,Set}` where necessary.

Please reach out on zulip if there are any questions.

**Comment by NiklasJonsson** [comment]
It seems like no one is assigned to this and I'd like to give it a try so assigning myself.

@rustbot claim

**Comment by NiklasJonsson** [comment]
`compiler/rustc_session/src/config.rs` has this code that fails when enabling the lint:
```
impl<T> CheckCfg<T> {
    fn map_data<O: Eq + Hash>(&self, f: impl Fn(&T) -> O) -> CheckCfg<O> {
        CheckCfg {
            names_valid: self
                .names_valid // FxHashSet
                .as_ref()
                .map(|names_valid| names_valid.iter().map(|a| f(a)).collect()),
            values_valid: self
                .values_valid // FxHashMap<T, FxHashSet>
                .iter()
                .map(|(a, b)| (f(a), b.iter().map(|b| f(b)).collect()))
                .collect(),
            well_known_values: self.well_known_values,
        }
    }
}
```
which seems fine enough as it just creates a new hashmap in-place with the mapped values and keys, assuming `f` doesn't have side-effects, which I don't think it can as it is `Fn`. But we want to replace this, which should be straight-forward but the file also has this code:
```
            // Get all values map at once otherwise it would be costly.
            // (8 values * 220 targets ~= 1760 times, at the time of writing this comment).
            let [
                values_target_os,
                values_target_family,
                values_target_arch,
                values_target_endian,
                values_target_env,
                values_target_abi,
                values_target_vendor,
                values_target_pointer_width,
            ] = self
                .values_valid
                .get_many_mut(VALUES)
                .expect("unable to get all the check-cfg values buckets");
```
which makes use of `get_many_mut` which is only implemented on `HashMap` and not the `IndexMap`, which means we can't replace it to avoid the lint. I'll close this for now and see if I can get together something with a smaller scope.

**Comment by NiklasJonsson** [comment]
I was hoping to get back to this after vacation but I haven't had the time and I likely won't in the near future. I've converted some of the modules to use the stable hash containers but there is a decent amount left so it is probably possible to have multiple people working on this, assuming some coordination.

There are two outstanding PRs from me that I will finish up. The first is https://github.com/rust-lang/rust/pull/99334 and seems ready to merge. Second one is https://github.com/rust-lang/rust/pull/99065 which requires the `get_many_mut` feature to be merged for the stable hash containers https://github.com/bluss/indexmap/pull/238, which in turn is waiting on the stabilization of the matching features on the standard hash containers (https://github.com/rust-lang/rust/issues/97601).

@rustbot release-assignment

**Comment by CastilloDel** [comment]
I'm gonna give this a try
@rustbot claim

**Comment by ndrewxie** [comment]
I'd like to take a crack at this, and maybe knock a few more of those `#[allow(rustc::potential_query_instability)]`s off

Just a quick question, however - a change like this might have the potential for some performance regressions. What's the best way for me to detect these? I could always do local timing/benchmarking, but I use wsl so I'm not sure how accurate my results would be. Would switching to pure windows (no WSL) lead to more accurate measurements?

@rustbot claim

**Comment by jyn514** [comment]
@ndrewxie don't worry about hogging resources, we have plenty of capacity. There's a thread for requesting someone to run the perf bot: https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/perf.20run
and some instructions on local benchmarking here: https://rustc-dev-guide.rust-lang.org/profiling.html

**Comment by ndrewxie** [comment]
I'm currently working on another issue, and spring break is fast ending, so I'll un-assign myself from this issue for now, in case someone else wants to work on it.

I'll probably return to it next week (or some time this month) if nobody takes it up in the meantime, though.

@rustbot release-assignment

**Comment by chenyukang** [comment]
There may be more need to clean up, for anyone maybe interested can refer to https://github.com/rust-lang/rust/pull/120485

**Comment by ismailarilik** [comment]
This PR is for [`rustc_hir_analysis`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_hir_analysis/src/lib.rs): https://github.com/rust-lang/rust/pull/131140

**Comment by ismailarilik** [comment]
This PR is for [`rustc_metadata`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_metadata/src/lib.rs): https://github.com/rust-lang/rust/pull/131145

**Comment by ismailarilik** [comment]
> The lint exists since [#89558](https://github.com/rust-lang/rust/pull/89558), and has allowed on all compiler crates as `rustc::potential_query_instability`. We now need to stop allowing it in the compiler.
> 
> Steps, for each crate except `rustc_data_structures`:
> 
>     * remove the `allow(rustc::potential_query_instability)` in `lib.rs`;
> 
>     * replace the `FxHash{Map,Set}` by `FxIndex{Map,Set}` where necessary.
> 
> 
> Please reach out on zulip if there are any questions.

@cjgillot You specified `rustc_data_structures` as an exception here. So what should be done for it? Is it nothing or another approach?

BTW, with the PRs I recently opened, there is not much left to fix but I will open another PR for the remainings.

**Comment by cjgillot** [comment]
`rustc_data_structures` is a small crate with support code. The lint is meant to avoid accidentally introducing non-determinism in compiler code.

The only thing that would be needed for `rustc_data_structures` is to review all the places where there is an `#[allow(rustc::potential_query_instability)]` and put a comment there explaining why it's ok.

**Comment by ismailarilik** [comment]
> `rustc_data_structures` is a small crate with support code. The lint is meant to avoid accidentally introducing non-determinism in compiler code.
> 
> The only thing that would be needed for `rustc_data_structures` is to review all the places where there is an `#[allow(rustc::potential_query_instability)]` and put a comment there explaining why it's ok.

@cjgillot there is just one `#[allow(rustc::potential_query_instability)]` for `rustc_data_structures` in [`lib.rs`](https://github.com/rust-lang/rust/blob/master/compiler/rustc_data_structures/src/lib.rs). Is your statement here is enough for rationale explanation? (I was thinking to add it as a comment above the allow statement.)

**Comment by ismailarilik** [comment]
For migrated ones, perf might be improved by removing some `sort` calls which were only added to provide order stability but I don't know how I can find them easily.

**Comment by fee1-dead** [comment]
@Aditya-PS-05 Are you still working on this?

**Comment by ismailarilik** [comment]
I think https://github.com/rust-lang/rust/pull/131200 can close this PR since other usages of `allow(rustc::potential_query_instability)` are either intentional or OK. Could you check as well; I might miss somethings.
